llvm/lib/SYCLLowerIR/SYCLOptimizeBarriers.cpp

```diff
@@ -0,0 +1,979 @@
+      if ((CmpExec == CompareRes::BIGGER || CmpMem == CompareRes::BIGGER) &&
+          semanticsSuperset(LastSem, CurSem) &&
+          noFencedMemAccessesBetween(Last.CI, Cur.CI, FenceCur, BBMemInfo)) {
+        if (MergedSem != LastSem) {
+          Last.CI->setArgOperand(2, ConstantInt::get(Int32Ty, MergedSem));
+          Last.Semantic = MergedSem;
+        }
+        Changed |= eraseBarrierWithITT(Cur);
+        break;
+      }
+      if ((CmpExec == CompareRes::SMALLER || CmpMem == CompareRes::SMALLER) &&
+          semanticsSuperset(CurSem, LastSem) &&
+          noFencedMemAccessesBetween(Last.CI, Cur.CI, FenceLast, BBMemInfo)) {
+        if (MergedSem != CurSem) {
+          Cur.CI->setArgOperand(2, ConstantInt::get(Int32Ty, MergedSem));
+          Cur.Semantic = MergedSem;
+        }
+        Changed |= eraseBarrierWithITT(Last);
+        Survivors.pop_back();
+        continue;
+      }
+      break;
+    }
+    if (Cur.CI) // still alive?
+      Survivors.emplace_back(Cur);
+  }
+
+  // If we removed any, replace Barriers with the survivors
+  if (Survivors.size() != Barriers.size()) {
+    Barriers.clear();
+    Barriers.append(Survivors.begin(), Survivors.end());
+    Changed = true;
+  }
+  return Changed;
+}
+
+// Walk the whole CFG once, first trying to erase fully–redundant
+// barriers and, if that is impossible, trying to downgrade
+// Cross-work-group barriers that are safely covered by another global fence.
+static bool optimizeBarriersCFG(SmallVectorImpl<BarrierDesc *> &Barriers,
+                                DominatorTree &DT, PostDominatorTree &PDT,
+                                BBMemInfoMap &BBMemInfo) {
+  bool Changed = false;
+
+  for (BarrierDesc *B : Barriers) {
+    if (!B->CI)
+      continue; // Already removed
+
+    bool Removed = false;
+    bool IsGlobalB =
+        (B->MemScope == Scope::Device || B->MemScope == Scope::CrossDevice ||
+         (B->Semantic &
+          static_cast<uint32_t>(MemorySemantics::CrossWorkgroupMemory)));
+    BarrierDesc *DowngradeCand = nullptr;
+
+    for (BarrierDesc *A : Barriers) {
+      if (A == B || !A->CI)
+        continue;
+
+      // Elimination check.
+      auto ExecCmp = compareScopesWithWeights(A->ExecScope, B->ExecScope);
+      auto MemCmp = compareScopesWithWeights(A->MemScope, B->MemScope);
+      bool ScopesCover =
+          (ExecCmp == CompareRes::BIGGER || ExecCmp == CompareRes::EQUAL) &&
+          (MemCmp == CompareRes::BIGGER || MemCmp == CompareRes::EQUAL);
+      bool SemCover = (A->Semantic & B->Semantic) == B->Semantic;
+      bool ADominatesB = DT.dominates(A->CI, B->CI);
+      if (ScopesCover && SemCover) {
+        RegionMemScope Fence = getBarrierFencedScope(*A);
+        // FIXME: this check is way too conservative.
+        if (Fence != RegionMemScope::Unknown && ADominatesB &&
+            PDT.dominates(B->CI, A->CI) &&
+            noFencedAccessesCFG(A->CI, B->CI, Fence, BBMemInfo)) {
```

Comment:
I think there could be repeated calculation if this function is called multiple times and a basic block is re-scanned in multiple ranges (A, B), (A, C), etc.
It might be enough to calculate the highest ExecScope and MemScope score of each basic block only once. Then the score can be used if the basic block is scanned multiple times.

Reply:
Correct, do you mean on [line 532](https://github.com/intel/llvm/pull/19353/files#diff-88d6917b6c0b3368d1e3ce74c14c6a6236bf2a08e7117df985a42d737af4867eR532) just compare with Required? Realistically it should be enough, though in some rare scenarios it will result in a missing optimization. For now I've added a shortcut in hasFencedAccesses for cases when path from barrier A to barrier B is looking like this: BB-A -> other BB -> BB-B - with this shortcut scan for other BB will be omitted.

Reply:
I mean each basic block has its own score, i.e. highest score of the instructions within the basic block. After calculating the score, there is no need to iterate over instructions over and over again.

Reply:
I've added an early exist for the required scope vs highest memory scope in BB. I'm hesitating to add per-instruction caching to avoid 100% of the cases for memory scope recalculation. TBH for this the pass should be re-written with a new type of metadata introduced.

Reply:
I see you already reused in `BBMemInfo` for `hasFencedAccesses` in commit https://github.com/intel/llvm/pull/19353/commits/cd8096b57c6b0695ec4b1c41248c481eedfb6e68.
A follow-up question is, can we unconditionally use `BBMemInfo` for checking and don't run `classifyMemScope` on instructions between `Start` and `End` in `hasFencedAccesses`? The benefit would be less calculation. My understanding is that it may misses optimization at a few case but won't cause incorrectness issue.

Reply:
Actually, I'm not able to get IR with this reduction block merged with its predecessor. Let me do few more experiments and probably can remove these extra scans

Reply:
Simplified the CFG walk, added an early exit, yet there is a potential rescan, but overall number of rescans were reduced. Some optimizations are now missing, but probably for good. Main optimization still works bringing performance gains.

Reply:
thanks @MrSidims